### DISCUSSION TITLE
How to start an https server when we build the app in standalone mode?

### DISCUSSION THREAD

**Discussion by thangkieu** [vercel/next.js#63659] [App Router] [answered]
I understand that https server feature is only available in development mode.
Do you guys have a work around to start a https server for standalone build? 
Thanks

**Comment by SouthLink** [comment]
standalone mode outputs a minimal server.js file that is used to launch the next.js service. See the example Dockerfile below. Start server.js with node.js or pm2
[with-docker](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile)

**Reply by thangkieu** [reply]
Thanks for your reply.
My issue is I cannot start a https server with `node server.js` inside the docker because `server.js` only supports http server

**Reply by SouthLink** [reply, marked as answer]
I understand that this should have nothing to do with standalone mode, you should need nginx to do this

**Reply by thangkieu** [reply]
Thanks

**Comment by jaj232** [comment]
@SouthLink is there no way to do this without nginx/Docker?  If I have a `server.js` file like this in the root directory:
```
const https = require('https');
const fs = require('fs');
const { parse } = require('url');
const next = require('next');

const dev = process.env.NODE_ENV !== 'production';
const app = next({});
const handle = app.getRequestHandler();

const httpsOptions = {
    key: fs.readFileSync('server-key.pem'), // Correct path to your key
    cert: fs.readFileSync('server-cert.pem') // Correct path to your certificate
};

app.prepare().then(() => {

    https.createServer(httpsOptions, (req, res) => {
        const parsedUrl = parse(req.url, true);
        handle(req, res, parsedUrl);
    }).listen(3000, (err) => {
        if (err) throw err;
        console.log('> Ready on https://localhost:3000');
    });
    
});
```

and I run `node server.js` and I go to `https://localhost:3000` it works.  How to adjust `standalone/server.js` file so that it works this way?

**Reply by SouthLink** [reply]
> @SouthLink is there no way to do this without nginx/Docker? If I have a `server.js` file like this in the root directory:
> 
> ```
> const https = require('https');
> const fs = require('fs');
> const { parse } = require('url');
> const next = require('next');
> 
> const dev = process.env.NODE_ENV !== 'production';
> const app = next({});
> const handle = app.getRequestHandler();
> 
> const httpsOptions = {
>     key: fs.readFileSync('server-key.pem'), // Correct path to your key
>     cert: fs.readFileSync('server-cert.pem') // Correct path to your certificate
> };
> 
> app.prepare().then(() => {
> 
>     https.createServer(httpsOptions, (req, res) => {
>         const parsedUrl = parse(req.url, true);
>         handle(req, res, parsedUrl);
>     }).listen(3000, (err) => {
>         if (err) throw err;
>         console.log('> Ready on https://localhost:3000');
>     });
>     
> });
> ```
> 
> and I run `node server.js` and I go to `https://localhost:3000` it works. How to adjust `standalone/server.js` file so that it works this way?

It looks like you have a custom service file, but custom services are not supported in standalone mode. If you don't want to use docker, you can use bash commands to implement similar features, but the runtime environment needs to be configured by yourself, but if you want to deploy to a service like aws in production mode, you should need nginx to do it

> When using standalone output mode, it does not trace custom server files. This mode outputs a separate minimal server.js file, instead. These cannot be used together.

**Comment by Calebjoshua** [comment]
Hi @SouthLink , if i use ALB in AWS and if i configure my node.js application in standalone mode. I also have a dockerfile. But now how to start my application with https protocol ?
